pkg/ccl/sqlproxyccl/balancer/assignment.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package balancer
+
+import (
+	"sync"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+)
+
+// ServerAssignment represents an assignment to a SQL pod.
+type ServerAssignment struct {
+	owner   ConnectionHandle
+	addr    string
+	onClose struct {
+		sync.Once
+		closerFn func()
+	}
+}
+
+// NewServerAssignment returns a new instance of ServerAssignment. Instances of
+// ServerAssignment must be closed through Close once no longer in use.
+func NewServerAssignment(
+	tenantID roachpb.TenantID, tracker *ConnTracker, owner ConnectionHandle, addr string,
+) *ServerAssignment {
+	sa := &ServerAssignment{owner: owner, addr: addr}
+	sa.onClose.closerFn = func() {
+		tracker.unregisterAssignment(tenantID, sa)
+	}
+	tracker.registerAssignment(tenantID, sa)
+	return sa
+}
+
+// Owner returns the connection handle associated with the server assignment.
+// The connection handle may not be initialized yet, so callers will need to
+// check for that where necessary.
+func (sa *ServerAssignment) Owner() ConnectionHandle {
+	return sa.owner
+}
+
+// Addr returns the address of the server assignment.
+func (sa *ServerAssignment) Addr() string {
+	return sa.addr
+}
+
+// Close cleans up the server assignment and deregisters it from the connection
+// tracker. This is idempotent.
+func (sa *ServerAssignment) Close() {
+	sa.onClose.Do(sa.onClose.closerFn)
```

Comment:
What do you think of using atomics here instead of sync.Once?

```
if atomics.CompareAndSwapUint32(&sa.onClose.done, 0, 1) && sa.onClose.closerFn != nil {
    sa.onclose.closerFun()
}
```

Personally, I'm torn on what is better. sync.Once is nice because it is in the standard library and it is easy to explain. The downside is there is actually a subtle deadlock risk. If the Close callback somehow leads to calling Close again, then sync.Once will deadlock. 

Using the atomic CAS avoids the deadlock risk. The downside of the atomics is it is a little unusual and it subtly changes the semantics of .Close(). With sync.Once, it is guaranteed that after .Close() returns, the clean up is complete. Using atomics means that if two Close() calls race, one may return before closerFn finishes running.

I also have two competing personal guidelines here:
1. Avoid the use of atomics for control flow. This is a comparatively simple use of atomics since the state transition is one way and isolated to a single function.
2. Avoid calling methods on another object while holding a lock.